Revised local state docs (#4378)
apollographql/apollo-client: docs/source/essentials/local-state.md

import { InMemoryCache } from 'apollo-cache-inmemory';

const client = new ApolloClient({
  cache: new InMemoryCache(),
  resolvers: { ... },
  typeDefs: { ... },
});
```

<dl>
  <dt>`resolvers?`: Resolvers | Resolvers[]</dt>
  <dd>A map of resolver functions that your GraphQL queries and mutations call in order to read and write to the cache.</dd>
  <dt>`typeDefs?`: string | string[] | DocumentNode | DocumentNode[];<string></dt>
  <dd>A string representing your client-side schema written in the [Schema Definition Language](/docs/graphql-tools/generate-schema.html#schema-language). This schema is not used for validation, but is used for introspection by the [Apollo Client Devtools](https://github.com/apollographql/apollo-client-devtools).</dd>
</dl>

None of these options are required. If you don't specify anything, you will still be able to use the `@client` directive to query the Apollo Client cache.

<h4 id="apollo-client-methods">Methods</h4>

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink } from 'apollo-link';

const client = new ApolloClient({
  cache: new InMemoryCache(),

>   link: ApolloLink.empty(),

Author: Good point - I'll switch it to a `HttpLink`. Thanks!

});

client.setResolvers({ ... });
```

<dl>
  <dt>`addResolvers(resolvers: Resolvers | Resolvers[])`</dt>
  <dd>A map of resolver functions that your GraphQL queries and mutations call in order to read and write to the cache. Resolver functions added through `addResolvers` are added to the internal resolver function map, meaning any existing resolvers (that aren't overwritten) are preserved.</dd>
  <dt>`setResolvers(resolvers: Resolvers | Resolvers[])`: </dt>
  <dd>A map of resolver functions that your GraphQL queries and mutations call in order to read and write to the cache. Resolver functions added through `setResolvers` overwrite all existing resolvers (a pre-existing resolver map is wiped out, before the new resolvers are added).</dd>
  <dt>`getResolvers`</dt>
  <dd>Get the currently defined resolver map.</dd>

>   <dt>`setTypeDefs(typeDefs: string | string[] | DocumentNode | DocumentNode[])`</dt>

Author: We've removed the `typeDefs` getters/setters (to help avoid this confusion). Thanks!

  <dd>A string representing your client-side schema written in the [Schema Definition Language](/docs/graphql-tools/generate-schema.html#schema-language), or as a GraphQL AST. This schema is not used for validation, but is used for introspection by the [Apollo Client Devtools](https://github.com/apollographql/apollo-client-devtools).</dd>
  <dt>`getTypeDefs`</dt>

>   <dd>Returns a string (SDL) or GraphQL AST representation of your client-side schema (if previously set via the `ApolloClient` constructor or `setTypeDefs`).</dd>

Author: We've talked about this separately, and have decided to remove the `typeDefs` getters/setters. For history's sake, they were included to make it easier to get `typeDefs` in Apollo Client Devtools. We've since decided to let devtools access the private `typeDefs` member (for now, since devtools aren't using TS). Thanks!

  <dt>`setLocalStateFragmentMatcher(fragmentMatcher: FragmentMatcher)`</dt>
  <dd>Set a custom `FragmentMatcher` to be used when resolving local state queries involving [fragments on unions or interfaces](/docs/react/advanced/fragments.html#fragment-matcher).</dd>
</dl>

**Typescript interfaces/types:**

```ts
interface Resolvers {
  [key: string]: {
    [field: string]: (
      rootValue?: any,
      args?: any,
      context?: any,
      info?: any,
    ) => any;
  };
}

type FragmentMatcher = (
  rootValue: any,
  typeCondition: string,
  context: any,
) => boolean;
```

<h3 id="apollo-cache">ApolloCache</h3>

<h4 id="apollo-cache-methods">Methods</h4>

```js
import { InMemoryCache } from 'apollo-cache-inmemory';

> cache.writeData({

Author: Let's show it being created and keep the import. We've accepted a lot of AC docs PR's that add imports and other details, to make it more clear what's happening, so people definitely like this extra detail. Thanks!

  data: {
    isLoggedIn: !!localStorage.getItem('token'),
    cartItems: [],
  },
});
```

<dl>
  <dt>`writeData({ id, data })`</dt>
  <dd>Write data directly to the cache without having to pass in a query. Great for prepping the cache with initial data.</dd>
  <dt>`writeQuery({ query, variables, data })`</dt>
  <dd>Similar to `writeData` (writes data to the cache) but uses the specified query to validate that the shape of the data you’re writing to the cache is the same as the shape of the data required by the query.</dd>
  <dt>`readQuery({ query, variables })`</dt>
  <dd>Read data from the cache for the specified query.</dd>
  <dt>`writeFragment({ id, fragment, fragmentName, variables, data })`</dt>
  <dd>Similar to `writeData` (writes data to the cache) but uses the specified fragment to validate that the shape of the data you’re writing to the cache is the same as the shape of the data required by the fragment.</dd>
  <dt>`readFragment({ id, fragment, fragmentName, variables })`</dt>
  <dd>Read data from the cache for the specified fragment.</dd>
</dl>

<h2 id="updating-local-state">Updating local state</h2>

> There are two main ways to perform local state mutations. The first way is to directly write to the cache by calling `cache.writeData` within an `ApolloConsumer` or through a `Query` component. Direct writes are great for one-off mutations that don't depend on the data that's currently in the cache, such as writing a single value. The second way is by creating a `Mutation` component with a GraphQL mutation that calls a local client-side resolver. We recommend using resolvers if your mutation depends on existing values in the cache, such as adding an item to a list or toggling a boolean. You can think of direct writes like calling React's `setState`, whereas local resolvers offer a bit more structure like Redux.

Author: Great revision - and great idea @peggyrayzis; I'll make an FR note of that!

To help make local resolvers more flexible, the signature of a resolver function is the exact same as resolver functions on the server built with [`graphql-tools`](/docs/graphql-tools/resolvers.html#Resolver-function-signature). Let's recap the four parameters of a resolver function:

```js
fieldName: (obj, args, context, info) => result;
```

1. `obj`: The object containing the result returned from the resolver on the parent field or the `ROOT_QUERY` object in the case of a top-level query or mutation.
2. `args`: An object containing all of the arguments passed into the field. For example, if you called a mutation with `updateNetworkStatus(isConnected: true)`, the `args` object would be `{ isConnected: true }`.

> 3. `context`: The context object, which is shared between your React components and your Apollo Client network stack. The most important thing to note here is that we've added the Apollo cache to the context for you, so you can manipulate the cache with `readQuery`, `writeQuery`, `readFragment`, `writeFragment`, and `writeData`. You can learn more about these methods in [Managing the cache](#managing-the-cache). We've also added the current Apollo Client instance to the `context` under a `client` property, in-case you're interested in calling into other parts of the Apollo Client API from a resolver.

Author: Great revision @benjamn, thanks! 

@peggyrayzis things can definitely get a bit ugly by allowing the full `client` instance to be available. We went back and forth on this, and decided to allow it since this feature was so highly requested. If people call `clearStore/resetStore` (and they don't handle resetting default values themselves), they can definitely cause problems. Good idea on the bulleted list, and maybe I'll add a small message here about being careful when calling parts of the `client` API.

> ⚠️ Since the above query runs as soon as the component is mounted, what do we do if there are no todos in the cache or there aren't any local resolvers defined to help calculate `todos`? We need to write an initial state to the cache before the query is run to prevent it from erroring out. Refer to the [Initializing the cache](#cache-initialization) section above for more information.

<h3 id="cache-initialization">Initializing the cache</h3>

Often, you'll need to write an initial state to the cache so any components querying data before a mutation is triggered don't error out. To accomplish this, you can use `cache.writeData` to prep the cache with initial values. The shape of your initial state should match how you plan to query it in your application.

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';

const cache = new InMemoryCache();
const client = new ApolloClient({
  cache,
  resolvers: { /* ... */ },
});

cache.writeData({
  data: {
    todos: [],
    visibilityFilter: 'SHOW_ALL',
    networkStatus: {
      __typename: 'NetworkStatus',
      isConnected: false,
    },
  },
});
```

> Sometimes you may need to [reset the store](/docs/react/features/cache-updates.html#reset-store) in your application, when a user logs out for example. If you call `client.resetStore` anywhere in your application, you will likely want to initialize your cache again. You can do this using the `client.onResetStore` method to register a callback that will call `cache.writeData` again.

Author: Great idea - I'll add a code snippet.

  return cartItems.includes(launch.id);
},
```

`launch` holds the data returned from the server for the rest of the query, which means in this case we can use `launch` to get the current launch `id`. We aren't using any arguments in this resolver, so we can skip the second resolver parameter. From the `context` however (the third parameter), we're using the `cache` reference, to work directly with the cache ourselves. So in this resolver, we're making a call directly to the cache to get all cart items, checking to see if any of those loaded cart items matches the parent  `launch.id`, and returning `true` / `false` accordingly. The returned boolean is then incorporated back into the result of running the original query.

Just like resolvers on the server, local resolvers are extremely flexible. They can be used to perform any kind of local computation you want, before returning a result for the specified field. You can manually query (or write to) the cache in different ways, call other helper utilities or libraries to prep/validate/clean data, track statistics, call into other data stores to prep a result, etc.

<h4 id="async-resolvers">Async local resolvers</h4>

> Apollo Client supports asynchronous local resolver functions. These functions can either be `async` functions or ordinary functions that return a Promise. This can be useful for performing side effects like accessing a device API.

Reviewer: ```suggestion
Apollo Client supports asynchronous local resolver functions. These functions can either be `async` functions or ordinary functions that return a `Promise`. Asynchronous resolvers are useful when they need to return data from an asynchronous API.
```

> ⚠️ If you would like to hit a REST endpoint from your resolver, [we recommend checking out `apollo-link-rest`](https://github.com/apollographql/apollo-link-rest) instead, which is a more complete solution for using REST endpoints with Apollo Client.

> For React Native and most browser APIs, you should set up a listener in a component lifecycle method and pass in your mutation trigger function as the callback instead of using an async resolver. However, there are some cases where it's beneficial to perform a side effect within a resolver:

Reviewer: ```suggestion
For React Native and most browser APIs, you should set up a listener in a component lifecycle method and pass in your mutation trigger function as the callback instead of using an async resolver. However, an `async` resolver function is often the most convenient way to consume asynchronous device APIs:
```

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { CameraRoll } from 'react-native';

const client = new ApolloClient({
  cache: new InMemoryCache(),
  resolvers: {
    Query: {

>       cameraRoll: async (_, { assetType }) => {

Reviewer: ```suggestion
      async cameraRoll(_, { assetType }) {
```

        try {
          const media = await CameraRoll.getPhotos({
            first: 20,
            assetType,
          });

          return {
            ...media,
            id: assetType,
            __typename: 'CameraRoll',
          };
        } catch (e) {
          console.error(e);
          return null;
        }
      },
    },
  },
});
```

> [`CameraRoll.getPhotos()`](https://facebook.github.io/react-native/docs/cameraroll.html#getphotos) returns a Promise resolving to an object with a `edges` property, which is an array of camera node objects, and a `page_info` property, which is an object with pagination information. This is a great use case for GraphQL, since we can filter down the return value to only the data that our components consume.

Reviewer: ```suggestion
[`CameraRoll.getPhotos()`](https://facebook.github.io/react-native/docs/cameraroll.html#getphotos) returns a `Promise` resolving to an object with a `edges` property, which is an array of camera node objects, and a `page_info` property, which is an object with pagination information. This is a great use case for GraphQL, since we can filter down the return value to only the data that our components consume.
```

```js
import gql from 'graphql-tag';

const GET_PHOTOS = gql`
  query getPhotos($assetType: String!) {
    cameraRoll(assetType: $assetType) @client {
      id
      edges {
        node {
          image {
            uri
          }
          location {
            latitude
            longitude
          }
        }
      }
    }
  }
`;
```

<h3 id="client-fields-cache">Handling `@client` fields with the cache</h3>

> As outlined in [Handling `@client` fields with resolvers](#client-fields-resolvers), `@client` fields can be resolved with the help of local resolver functions. While local resolvers can be super helpful, it's important to note that they are not always required when using an `@client` directive. Fields marked with `@client` can still be resolved locally, by pulling matching values out of the cache directly. Let's look at an example of this:

Reviewer: ```suggestion
As outlined in [Handling `@client` fields with resolvers](#client-fields-resolvers), `@client` fields can be resolved with the help of local resolver functions. However, it's important to note that local resolvers are not always required when using an `@client` directive. Fields marked with `@client` can still be resolved locally, by pulling matching values out of the cache directly. For example:
```

<h3 id="write-read-query">writeQuery and readQuery</h3>

Sometimes, the data you're writing to the cache depends on data that's already in the cache; for example, you're adding an item to a list or setting a property based on an existing property value. In that case, you should use `cache.readQuery` to pass in a query and read a value from the cache before you write any data. Let's look at an example where we add a todo to a list:

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import gql from 'graphql-tag';

let nextTodoId = 0;

const cache = new InMemoryCache();
cache.writeData({
  data: {
    todos: [],
  },
});

const client = new ApolloClient({
  resolvers: {
    Mutation: {
      addTodo: (_, { text }, { cache }) => {
        const query = gql`
          query GetTodos {
            todos @client {
              id
              text
              completed
            }
          }
        `;

        const previous = cache.readQuery({ query });
        const newTodo = { id: nextTodoId++, text, completed: false, __typename: 'TodoItem' };
        const data = {

>           todos: previous.todos.concat([newTodo]),

Reviewer: ```suggestion
          todos: [...previous.todos, newTodo],
```

The following demonstrates how to configure a client-side schema through the `ApolloClient` constructor:

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink } from 'apollo-link';
import gql from 'graphql-tag';

const typeDefs = gql`
  extend type Query {
    isLoggedIn: Boolean!
    cartItems: [Launch]!
  }

  extend type Launch {
    isInCart: Boolean!
  }

  extend type Mutation {
    addOrRemoveFromCart(id: ID!): [Launch]
  }
`;

const client = new ApolloClient({
  cache: new InMemoryCache(),
  link: ApolloLink.empty(),
  typeDefs,
});
```

If you open up Apollo Client Devtools and click on the `GraphiQL` tab, you'll be able to explore your client schema in the "Docs" section. This example doesn't include a remote schema, but if it did, you would be able to see your local queries and mutations alongside your remote ones.

> ![GraphiQL Console](../img/client-schema.png)

Author: For some reason this image isn't showing properly in the Netlify preview. Hexo was corrupting the file, but when I moved it to a new directory and tested locally (with a fresh Hexo cache), that seems to have fixed the issue. I'm guessing the deploy preview is just in a bit of a broken state with regards to the image, but I'll look into this further.

Just to add - this image is also currently broken in production (https://www.apollographql.com/docs/react/essentials/local-state.html#schema). So Hexo has been corrupting the image for a while ... 🤔

Reviewer 2: @hwillson I think it accidentally broke during the LF conversion, in this commit: https://github.com/apollographql/apollo-client/pull/4232/commits/53d33eb99a02722ff7edb8e9ddbc3cc72e750afc

I could make an undo PR for just the `png`s if you'd like.

Reviewer 2: #4438

Author: Oh, wow @rajington - awesome find! Thanks for the PR!

Reviewer 2: absolutely! and great work on this project, glad I could help!

<h2 id="migrating">Migrating from `apollo-link-state`</h2>

The [`apollo-link-state`](https://github.com/apollographql/apollo-link-state) project was the first to bring local state handling into the Apollo ecosystem. Handling local resolvers through the addition of an `ApolloLink` was a great starting point, and proved that `@client` based queries make sense, and work really well for local state management.

> While `apollo-link-state` works well, and it's nice to be able to modularly include local state handling when required, the use of an `ApolloLink` introduces a few hard to work around limitations. Local state handling in a link is restricted to only being able to leverage the API elements available within the link. We see local state management as an important part of the Apollo ecosystem, and as Apollo Client progresses, we want to make sure the handling of local resolvers are integrated as tightly as possible into core. This opens up new possibilities (like `@export` handling) and will tie nicely into the future planned adjustments to cache data retention, invalidation, garbage collection, etc.

Reviewer: ```suggestion
While `apollo-link-state` achieved some of the goals of local state handling, the information available when using any `ApolloLink` is limited by the modularity of the link system. We consider local state management a core part of the Apollo ecosystem, and as Apollo Client progresses, we want to make sure local resolvers are integrated as tightly as possible into core. This integration opens up new possibilities (like `@export` handling) and ties nicely into the future planned adjustments to cache data retention, invalidation, garbage collection, and other planned features that impact both local and remote data.
```

Updating your application to use Apollo Client's local state management features, instead of `apollo-link-state`, is fairly straightforward. The necessary steps are outlined below.

1. Including `apollo-link-state` as a dependency, and importing it to use `withClientState`, is no longer necessary. You can remove the `apollo-link-state` dependency since local state management is included with `apollo-client` >= 2.5.0.

2. Using `withClientState` is no longer supported. The following

  ```js
  const cache = new InMemoryCache();
  const stateLink = withClientState({ cache, resolvers: { ... } });
  const link = ApolloLink.from([stateLink, new HttpLink({ uri: '...' })]);
  const client = new ApolloClient({
    cache,
    link,
  });
  ```
  becomes

  ```js
  const client = new ApolloClient({
    cache: new InMemoryCache(),
    link: new HttpLink({ uri: '...' }),
    resolvers: { ... },
  });
  ```

>   It's important to note here that `defaults` are no longer supported. To prep the cache, use [`cache.writeData`](#write-data) directly instead. So

Reviewer: Should this be a separate list item?

Author: Yes, I think so - great point!

);
```

Here we create our GraphQL query and add `@client` directives to `todos` and `visibilityFilter`. We then pass the query to our `Query` component. The `@client` directives here let the `Query` component know that `todos` and `visibilityFilter` should be pulled from the Apollo Client cache or resolved using pre-defined local resolvers. The following sections help explain how both options work in more detail.

> ⚠️ Since the above query runs as soon as the component is mounted, what do we do if there are no todos in the cache or there aren't any local resolvers defined to help calculate `todos`? We need to write an initial state to the cache before the query is run to prevent it from erroring out. Refer to the [Initializing the cache](#cache-initialization) section above for more information.

<h3 id="cache-initialization">Initializing the cache</h3>

> Often, you'll need to write an initial state to the cache so any components querying data before a mutation is triggered don't error out. To accomplish this, you can use `cache.writeData` to prep the cache with initial values. The shape of your initial state should match how you plan to query it in your application.

Reviewer 3: [This comment](https://github.com/apollographql/apollo-client/pull/4155#issuecomment-455174470) about async initializers cleared up one aspect of the new API which had caused me not-insignificant confusion.  Does the synchronous nature of constructor initializers bear mention here?

Author: Thanks for the feedback @jangerhofer. We've decided to remove initializers from Apollo Client, and direct people towards using the existing Apollo Cache API to prepare the cache (`cache.writeData` specifically).

    launch(id: $launchId) {
      isInCart @client
      site
      rocket {
        type
      }
    }
  }
`;

// ... run the query using client.query, a <Query /> component, etc.
```

Here when the `GET_LAUNCH_DETAILS` query is executed, Apollo Client looks for a local resolver associated with the `isInCart` field. Since we've defined a local resolver for the `isInCart` field in the `ApolloClient` constructor, it finds a resolver it can use. This resolver function is run, then the result is calculated and merged in with the rest of the query result (if a local resolver can't be found, Apollo Client will check the cache for a matching field - see [Local data query flow](#query-flow) for more details).

Setting resolvers through `ApolloClient`'s constructor `resolvers` parameter, or through its `setResolvers` / `addResolvers` methods, adds resolvers to Apollo Client's internal resolver map (refer to the [Local resolvers](#local-resolvers) section for more details concerning the resolver map). In the above example we added a  `isInCart` resolver, for the `Launch` GraphQL object type, to the resolver map. Let's look at the `isInCart` resolver function more closely:

```js

> isInCart: (launch, _args, { cache }) => {

Author: Great point and will do. For some reason I couldn't get line highlighting to work properly, so I left it out 🤔.

While `@client(always: true)` ensures that a local resolver is always fired, it's important to note that if a query is using a `fetchPolicy` that leverages the cache first (`cache-first`, `cache-and-network`, `cache-only`), the query is still attempted to be resolved from the cache first, before the local resolver is fired.    This happens because `@client(always: true)` use could be mixed with normal `@client` use in the same query, which means we want part of the query to adhere to the defined `fetchPolicy`. The benefit of this is that anything that can be loaded from the cache first is made available to your `@client(always: true)` resolver function, as its [first parameter](#local-resolvers). So even though you've used `@client(always: true)` to identify that you want to always run a specific resolver, within that resolver you can look at the loaded cache values for the query, and decide if you want to proceed with running the resolver.

> <h2 id="combine-local-remote">Combining local and remote data</h2>

Author: Hmmm - yes, it is a bit duplicative. I kind of like keeping the ideas a bit separate though, so we're showing that `@client` fields can be used without remote data, but I think we can merge these sections together. I'll get rid of the `Combining local and remote data` section completely, and move `Integrating @client into remote queries` up into the `Handling @client fields with resolvers` as a sub-section. I'll then move `Using @client fields as variables` up into the `Querying local state` section. Let's see how that looks. Thanks!

  Apollo Client currently only supports the use of the `@export` directive when mixed with the `@client` directive. It prepares `@export` variables by first running through an operation that has `@client @export` directives, extracting the specified `@export` variables, then attempting to resolve the value of those variables from the local cache or local resolvers. Once a map of variable names to local values is built up, that map is then used to populate the variables passed in when running the server based GraphQL query. The execution order of the server based GraphQL query is not impacted by `@export` use; the variables are prepped and organized before the server query runs, so the specification is being followed.

3. If you define multiple `@export` variables that use the same name, in a single operation, the value of the last `@export` variable will be used as the variable value moving forward. When this happens Apollo Client will log a warning message (dev only).

> <h2 id="managing-the-cache">Managing the cache</h2>

Author: I’m on the fence on this one. I definitely hear what you’re saying, but working with the cache is such an integral part of the AC local state experience, that I wanted to make sure people could find out how to use it, without jumping around the docs. If we remove the `Managing the cache` section, and just point people to https://www.apollographql.com/docs/react/advanced/caching.html#direct, are we having them jump around a bit too much? Maybe not … one things for sure though, we’ll need to add `cache.writeData` to the `Direct Cache Access` section, since it isn’t mentioned anywhere.

So just to confirm @peggyrayzis, when you say point people to the advanced caching guide, you mean point them to [`Direct Cache Access`](https://www.apollographql.com/docs/react/advanced/caching.html#direct), correct? If so I’ll add `cache.writeData` to that section, remove `Managing the cache`, and update all links. Thanks!

            completed
          }
        `;
        const todo = cache.readFragment({ fragment, id });
        const data = { ...todo, completed: !todo.completed };

        // you can also do cache.writeData({ data, id }) here if you prefer
        cache.writeFragment({ fragment, id, data });
        return null;
      },
    },
  },
});
```

In order to toggle our todo, we need the todo and its status from the cache, which is why we call `cache.readFragment` and pass in a fragment to retrieve it. The `id` we're passing into `cache.readFragment` refers to its cache key. If you're using the `InMemoryCache` and not overriding the `dataObjectFromId` config property, your cache key should be `__typename:id`.

To write the data to the cache, you can use either `cache.writeFragment` or `cache.writeData`. The only difference between the two is that `cache.writeFragment` requires that you pass in a fragment to validate that the shape of the data you're writing to the cache node is the same as the shape of the data required by the fragment. Under the hood, `cache.writeData` automatically constructs a fragment from the `data` object and `id` you pass in and calls `cache.writeFragment`.

> <h2 id="client-schema">Client-side schema</h2>

Author: Hmm - good question. I'll think about this one a bit longer ...

The following demonstrates how to configure a client-side schema through the `ApolloClient` constructor:

```js
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { ApolloLink } from 'apollo-link';
import gql from 'graphql-tag';

const typeDefs = gql`
  extend type Query {
    isLoggedIn: Boolean!
    cartItems: [Launch]!
  }

  extend type Launch {
    isInCart: Boolean!
  }

  extend type Mutation {
    addOrRemoveFromCart(id: ID!): [Launch]
  }
`;

const client = new ApolloClient({
  cache: new InMemoryCache(),

>   link: ApolloLink.empty(),

Author: For sure - thanks!
